### ISSUE TITLE
[Feature Request]  Support Android TV / non touch devices

### ISSUE THREAD

**Issue by peat80** [ImranR98/Obtainium#281]
Hello, i just discovered this nice app and thought it would be nice to have to update some apps from github, mainly on my androidtv device. Unfortunately it does not (completely) work on non touch devices it seems. It opens fine in landscape and i can navigate to all buttons perfectly fine with remote, but there is no reaction when i press the ok button. So it looks to me like this app could be made to work with minimal effort. 🤔

Would be really nice if non touch support could be added in an future release. 😎👍
Keep up the good work.

**Comment by ImranR98** [comment]
Not sure how simple this will be, but definitely something to look into.

**Comment by peat80** [comment]
https://www.dltlabs.com/blog/how-to-create-an-android-tv-app-using-flutter-362536

**Comment by PlasmaticRiver** [comment]
> Not sure how simple this will be, but definitely something to look into.

It should be pretty easy. The only issue with it working on Android TV/Fire Stick is that pressing the enter/ok button in the remote doesn't register as a button tap, but the navigation mostly works fine already. That's the only thing that needs fixing.

**Comment by KingLucius** [comment]
Currently stuggle with FireTV noy having a good availability of apps and i have to sideload everything.
I ended creating a custom fdroid repo and manually upload everything to it and using Droid-ify to install which has partially-works navigation for android tv so i use mouse-toggle app to emulate the mouse 😅

So this one i am so interested about.

**Comment by lorrding** [comment]
Any update on this ? would be really helpful to have access to GitHub repo via Android TV as it's sometimes difficult to access internal storage or a web browser

**Comment by PlasmaticRiver** [comment]
> Any update on this ? would be really helpful to have access to GitHub repo via Android TV as it's sometimes difficult to access internal storage or a web browser

Still waiting on the Dev to patch in ATV support. Just using it sideloaded for now with Mouse Toggle to interact with the UI.

**Comment by ImranR98** [comment]
Fixing in next release (issue was https://github.com/flutter/flutter/issues/106817)

**Comment by ImranR98** [comment]
https://github.com/ImranR98/Obtainium/releases/tag/v0.13.26-beta

**Comment by peat80** [comment]
Sadly this only works partially. It only works on the first level. Pressing OK does do nothing in opened popups and dialogue boxes. Also OK does nothing when you enter text and want to confirm your entry.

**Comment by WardPearce** [comment]
would be a great addition, as updating apps on android tv is always a pain

**Comment by peat80** [comment]
> would be a great addition, as updating apps on android tv is always a pain

You should try Apkupdater on your tv too.
A great tool that has a nice tv ui. 😀

https://github.com/rumboalla/apkupdater

**Comment by WardPearce** [comment]
thank you 🙌

**Comment by SecularSteve** [comment]
Currently testing it myself on a Chromecast w/ Google TV.
Need to open it through Aptoide because it has no banner.

Here's one on the house, made it myself in Inkscape, under the CC-BY SA 4.0.

[Obtainium_TV_banner.zip](https://github.com/ImranR98/Obtainium/files/13238647/Obtainium_TV_banner.zip)
![Obtainium_TV_banner](https://github.com/ImranR98/Obtainium/assets/33793273/bd4ff633-9b07-4820-b09a-17167d608064)
![Obtainium_TV_banner_padded](https://github.com/ImranR98/Obtainium/assets/33793273/75bdcd00-f623-4e52-a0b2-1d8270934df0)

**Comment by GenericMale** [comment]
This seems to have improved in a recent version (possibly by a flutter update?).
Confirming dialogs and navigating views seems to work fine now but the text boxes trap the focus in a way which makes adding apps impossible with just a remote and on-screen keyboard.

**Comment by peat80** [comment]
Of course not.
But it works a little better now.
If you configure your apps on the phone, make a backup and then import it on tv its atleast useable for updating them. 😎👌

**Comment by peat80** [comment]
To close the issue atleast leanback support with tv banner and working text boxes to edit and add apps on tv would be needed i think. 🤔

**Comment by GenericMale** [comment]
> So you think this issue could be closed?

Unfortunately no because it's still not possible to use the app on a TV with a remote.
I used an external keypad with a touchpad to add the apps but running updates works perfectly fine just with the TV remote (this wasn't working before).

The main thing which needs to be fixed are the text inputs.
Problem is that as soon as you focus a text field (e.g. using the D-Pad on the remote), the onscreen keyboard is shown and there is no way to close that and focus the next control. So you can't click the Add or Search Button or edit any of the other inputs because you get stuck at the first text field in the view. Clicking the enter key in the onscreen keyboard or pressing the back key on the remote will unfocus the control and you start traversing the controls from the top of the view again.

There are various workarounds proposed in https://github.com/flutter/flutter/issues/49783 but i couldn't get any of them to work.

One thing that kinda helps is to set `textInputAction: TextInputAction.next` on the TextFormField (in GeneratedFormSubForm). This will put the focus on the next control when the enter button is clicked in the onscreen keyboard. This behavior probably isn't obvious to users and it can be pretty annoying when you wanna go through a page with many text views (e.g. the Edit App view) but at least you can reach all the controls somehow without a touch device.

**Comment by cerels** [comment]
I manage to use obtanium by using a mouse app to navigate, the thing is background updates doesn't work at all even though the settings for it are turned on, is this a limitation of Google tv?

**Comment by DarkCrypt** [comment]
@ImranR98 
You can check this [flutter-movies](https://github.com/arafatamim/flutter-movies) repo and see how they handle this.

**Comment by tazztone** [comment]
wanted to install smarttube with it. somehow it broke my remote control navigation (buttons were no longer doing what they're supposed to do) and i had to reboot the tv box ( leap s3+). happened many times. not sure what about obtainium could have caused this

**Comment by peat80** [comment]
> I manage to use obtanium by using a mouse app to navigate, the thing is background updates doesn't work at all even though the settings for it are turned on, is this a limitation of Google tv?

Background updates require an os that supports that.
I don't know for sure but i think you need atleast android (tv) 12.

They did not work on my shield, but on my google tv streamer (android tv 14) obtainium does indeed background updates.
Only smarttube is not wanting to do that so far. 😀

**Comment by peat80** [comment]
Same here for me on gtv streamer.
However, importing bsckups works fine for me.

So i always manage and edit my applist for tv on the phone, then export, transfer to streamer and import there.
Works fine for me.
On the streamer with atv 14 even background updates  do work for some apps.
Some only update manually. 🤷‍♂️

At least it works manually for all apps i have configured.  👌

**Comment by bmwagner18** [comment]
> Another issue on TV is the import/export option. Recent versions of Android TV have no way of selecting folders, so it's impossible to set a location for backup. It needs a location set as default.

I also am running into this issue on Onn 4K streaming device on the most recent update of ATV. I saw somewhere that disabling/uninstalling Play Store can cause this but it still does not work after clean factory reset. I would love an import option that can import from a URL so I can just type in a direct URL to the JSON and import that way.

**Comment by MarkyMarkDE** [comment]
> Of course not. But it works a little better now. If you configure your apps on the phone, make a backup and then import it on tv its atleast useable for updating them. 😎👌

Does unfortunately not work for me, there is no file picker on waipu.tv Stick 😞

**Comment by MarkyMarkDE** [comment]
> > I manage to use obtanium by using a mouse app to navigate, the thing is background updates doesn't work at all even though the settings for it are turned on, is this a limitation of Google tv?
> 
> Background updates require an os that supports that. I don't know for sure but i think you need atleast android (tv) 12.
> 
> They did not work on my shield, but on my google tv streamer (android tv 14) obtainium does indeed background updates. Only smarttube is not wanting to do that so far. 😀

Yes, the same information I also have, automatically updates are on AndroidTV >= 12 possible

**Comment by PlasmaticRiver** [comment]
>Does unfortunately not work for me, there is no file picker on waipu.tv Stick (AndroidTV 11) 😞.
If there is any news about this, I would glad to test.
In summer I become a tester for AndroidTV 14 on waipu.tv Stick, would also test this again.

It won't work for setting an export directory, but if you install a file explorer like Mixplorer on ATV it will work as a file picker to import from somewhere else as long as you give it storage permissions.

I mentioned this in another post about the problem #2076

**Comment by MarkyMarkDE** [comment]
> > Does unfortunately not work for me, there is no file picker on waipu.tv Stick (AndroidTV 11) 😞.
> > If there is any news about this, I would glad to test.
> > In summer I become a tester for AndroidTV 14 on waipu.tv Stick, would also test this again.
> 
> It won't work for setting an export directory, but if you install a file explorer like Mixplorer on ATV it will work as a file picker to import from somewhere else as long as you give it storage permissions.
> 
> I mentioned this in another post about the problem [#2076](https://github.com/ImranR98/Obtainium/issues/2076)

Indeed you are right, with Dateimanager+ import also work like your description, but with my expensive Filemanager from MobiSoft it didn't work...

**Comment by IngwiePhoenix** [comment]
I just finished setting up a custom launcher and a few other nicities on my ATV Chromecast (that babyblue-colored dongle) and wanted to use Obtanium to install a few apps and manage my updates there - but I couldn't even click the "Search" button without using `scrcpy`. So I ended up writing the JSON obtained from the crowd-sourced list into the `/storage/emulated/0/Android/data/dev.imranr.obtainium.fdroid/files/app_data` directory through `adb shell`; but the file was not recognized.

What is the best way to add new stuff there? While `scrcpy` works, I would love to use the pre-created JSON blobs from the community if possible.

Thanks!

**Comment by peat80** [comment]
Configuring and adding apps still is not really possible on tv without a mouse.
It is possible with emulating a mouse.

An awesome tool to manage your android tv
is AtvTools.
It can take screenshots, install apps,
move files from phone to tv or reverse.
And it has a mouse emulator.

With this you can use obtainium completely on tv using the mouse emulator.

For a big app list the easiest way to manage is still to configure them on the phone > then export as file > copy your exported file to tv > and then import that backup on your tv > and then have fun while easily updating your tv apps.

😉👌

**Comment by SecularSteve** [comment]
Here's something interesting that may reinvigorate interest here: 

https://github.com/LineageOS/android_packages_apps_Catapult

![image0.webp](https://github.com/user-attachments/assets/572ac258-dd9b-486d-ba02-dcda7a1c93c8)

Apparently, Lineage got tired of Google's crap they were pulling recently, that includes going Chinese on their Verified TV boxes/Chromecasts.

So they forked their own Android TV launcher they named "Catapult", and removed all the clutter for their latest 23 update.

**Comment by PlasmaticRiver** [comment]
@ImranR98 if you do add more optimizations for Android TV, I would throw this graphic on as the banner.png inside the APK so it displays correctly on TV launchers.

<img width="1024" height="500" alt="Image" src="https://github.com/user-attachments/assets/fab3daa5-826c-46bc-b165-9014979b34e6" />

**Comment by WolfganP** [comment]
> Configuring and adding apps still is not really possible on tv without a mouse. It is possible with emulating a mouse.
> 
> An awesome tool to manage your android tv is AtvTools. It can take screenshots, install apps, move files from phone to tv or reverse. And it has a mouse emulator.
> 
> With this you can use obtainium completely on tv using the mouse emulator.
> 
> For a big app list the easiest way to manage is still to configure them on the phone > then export as file > copy your exported file to tv > and then import that backup on your tv > and then have fun while easily updating your tv apps.

The main issue is alpha input fields to trigger the onscreen keyboard but retaining focus after escaping the keyboard, so it is impossible to navigate certain screens or popus (ie the search source selection)

Maybe playing with the focus shifting to the next element after editing the alpha fields, or mapping a remote control button (ie page up/down) to manage element focus or simulate a TAB keypress would be a workable workaround for the time being.

**Comment by xmicro** [comment]
What's the current status on this? Is it supported yet or no?

**Comment by MarkyMarkDE** [comment]
> What's the current status on this? Is it supported yet or no?

if you don't see further comments at this topic, then i would mean, no  - not at current, but it can be a further plan further - don't know, this ticket isn't closed with comment "not possible, too complicated" or whatever, so the chance is given to get this support in future version  :wink:.

**Comment by PlasmaticRiver** [comment]
> What's the current status on this? Is it supported yet or no?

best workaround at the moment is using a mouse toggle to navigate the screens with text, otherwise it works

**Comment by ACR-Jeff** [comment]
Works on my Nvidia Shield using a iPazzport mini keyboard with a mouse touchpad

**Comment by xmicro** [comment]
> > What's the current status on this? Is it supported yet or no?
> 
> best workaround at the moment is using a mouse toggle to navigate the screens with text, otherwise it works

Don't have a mouse? Any way to emulate one with a phone or something? Or another app?

**Comment by peat80** [comment]
> > > What's the current status on this? Is it supported yet or no?
> > 
> > 
> > best workaround at the moment is using a mouse toggle to navigate the screens with text, otherwise it works
> 
> Don't have a mouse? Any way to emulate one with a phone or something? Or another app?

Use tvQuickActions on your device or do it from phone with atvTools.
Both great apps and must haves for me.

**Comment by xmicro** [comment]
> > > > What's the current status on this? Is it supported yet or no?
> > > 
> > > 
> > > best workaround at the moment is using a mouse toggle to navigate the screens with text, otherwise it works
> > 
> > 
> > Don't have a mouse? Any way to emulate one with a phone or something? Or another app?
> 
> Use tvQuickActions on your device or do it from phone with atvTools. Both great apps and must haves for me.

Tried TVQuickAction but it did not work for some reason. Ended up using MATVT though, which worked fine.

**Comment by krgauthi** [comment]
Hey added basic functionality for android tv - mostly text field focus handling and autofocus on dialogs allows you to search and install apps pretty via the remote no keyboard / mouse workarounds needed!

https://github.com/ImranR98/Obtainium/pull/2831

**Comment by peat80** [comment]
Awesome, that sounds great.
Is there an apk to test for Download?

**Comment by krgauthi** [comment]
Here’s a debug APK to test with

https://github.com/krgauthi/Obtainium/releases

**Comment by peat80** [comment]
Thanks, have installed it.

Sadly it seems to be missing the Obtainium--Import and Obtainium-Export features under import / export.

Was this removed intentionally?
Can they added back please?
Importing a backup on tv works fine for the normal version of Obtainium.
Export also works if you install
Ultimate File Manager on your tv.
This filemanager has a proper documenzs provider which gets this function working on tvs.

So please add this function back in. 😉
Then i could do a proper test. 😀

Otherwise the app banner shows and it seems to work. 🥳

**Comment by krgauthi** [comment]
Interesting I didn't touch that behavior with these changes, like you mentioned you would need a file manager / web browser app installed for these intents to still function.

**Comment by peat80** [comment]
This seems to be a bug then.
Out of curiosity i tried to install this apk on my phone.
There it shows an overflow warning on the import / export tab if you tilt the phone to landscape layout.
So maybe the elements in this section simply do not fit in the tv layout and are hidden because of this? 🤔

<img width="2400" height="1080" alt="Image" src="https://github.com/user-attachments/assets/68eb0c8a-f57e-418d-93c3-51672662ce9d" />

**Comment by peat80** [comment]
So i played a little with this build, and overall it is a great improvement for tv. 👍

Adding apps via remote only works fine now.
Text fields focus nicely. 👌🏻

Have only some little issues / suggestions:

The missing Obtainium Import /Export options should be fixed. Must have functions for me to save my app list and settings.

On the main screen the little blue update icon can still not be reached via remote. This should be improved.
Also it would be cool to get the app icons back on the app list. They were probably replaced by the useful checkboxes. But it should be enough space there to show both.

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/57615be8-f3f8-4e9e-864b-660d35690e14" />

On the app page now everything is reachable, but the categories button is under the bottom sheet but reachable.
It takes too many clicks to navigate to the bottom sheet.
Solution here could be to either make it a top sheet / toolbar instead or just automatically focus a toolbar button, maybe the edit button when opening the page.

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/3cffde12-7a84-4e6e-858f-8b97d370f065" />

And the app is a little slow. But this is probably just the debug build. 😀

I hope this pr could be polished and find its way into a regular release soon.

Thx for the nice work krgauthi.
This is great. ♥️
